### PR TITLE
*: replace ddl callback injection with failpoint

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve?

Issue Number: ref #54436

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [x] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/server/handler/tikvhandler/tikv_handler.go [hunk 1]
```diff
@@ -1995,33 +1984,6 @@ func (h *TestHandler) handleGCResolveLocks(w http.ResponseWriter, req *http.Requ
 	}
 }
 
-// ServeHTTP handles request of resigning ddl owner.
-func (h DDLHookHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
```
Comment (D3Hunter) [inline pkg/server/handler/tikvhandler/tikv_handler.go:1999]: @AilinKid told me that it's used in other tests

Comment (lance6716) [reply pkg/server/handler/tikvhandler/tikv_handler.go:1999]: I found it's https://github.com/PingCAP-QE/automated-tests/blob/44e476617b26efdf344731d2d38393bbd8d76f1c/ticases/column_type_change/ctc_correctness_check/ctc_correctness_check.go#L51 

I will try to use add a failpoint and use `/fail/` TiDB HTTP API to replace it.

**Review by tangenta — APPROVED** [review]

File: pkg/server/handler/tikvhandler/tikv_handler.go (hunk 1 above)
Comment (D3Hunter) [reply pkg/server/handler/tikvhandler/tikv_handler.go:1999]: lgtm, maybe add a common api to enable failpoint dynamically, and better implement it at failpoint repo which i am not sure whether it exists

File: pkg/ddl/job_worker.go [hunk 2]
```diff
@@ -519,9 +519,19 @@ func (w *worker) transitOneJobStep(d *ddlCtx, job *model.Job) (int64, error) {
 		})
 		return 0, w.handleJobDone(d, job, t)
 	}
-	d.mu.RLock()
-	d.mu.hook.OnJobRunBefore(job)
-	d.mu.RUnlock()
+	failpoint.InjectCall("onJobRunBefore", job)
+	// ctcSlowDown is used by failpoint HTTP API in other repo, don't delete it.
+	failpoint.Inject("ctcSlowDown", func() {
```
Comment (D3Hunter) [inline pkg/ddl/job_worker.go:524]: can we enable it with EnableCall("onJobRunBefore") in the api

**Review by D3Hunter — COMMENTED** [review]
rest lgtm

File: pkg/server/handler/tikvhandler/tikv_handler.go (hunk 1 above)
Comment (lance6716) [reply pkg/server/handler/tikvhandler/tikv_handler.go:1999]: already there

https://github.com/pingcap/tidb/blob/3dfa15cd978624b10ea74931dcf171cc8dc6c8bb/pkg/server/http_status.go#L422-L430

File: pkg/ddl/job_worker.go (hunk 2 above)
Comment (lance6716) [reply pkg/ddl/job_worker.go:524]: `EnableCall` can only be enabled with code, not environment variables or HTTP API.

Comment (D3Hunter) [reply pkg/ddl/job_worker.go:524]: for this failpoint the logic is fixed, we can handle it differently at the http api side, not here, to make it cleaner

Comment (lance6716) [reply pkg/ddl/job_worker.go:524]: OK, I'll try to let `/fail/` HTTP handler call `EnableCall`

Comment (lance6716) [reply pkg/ddl/job_worker.go]: https://github.com/pingcap/tidb/pull/55215/commits/109c404b7e85d70c7efa239ab74223b341d9f4ba

**Review by D3Hunter — APPROVED** [review]

**Pull request merged**